Add PARALLEL support

Algunenano opened the pull request:
Add PARALLEL and VOLATILE tags for Postgresql functions

Algunenano commented:
After some final fixes I think this should be done with the PR.

Algunenano commented:
Solves https://github.com/CartoDB/crankshaft/issues/181

rafatower commented on src/pg/sql/cdb_utils.sql:
```diff
@@ -299,7 +299,7 @@ BEGIN
 
     RETURN (best_result)[2:array_upper(best_result, 1)];
 END;
-$$ language plpgsql IMMUTABLE;
```
Are you sure this is `VOLATILE`? the code seems to at least try to provide reproducibility (`setseed` above). Any side effects maybe?

Algunenano replied:
Technically both setseed and random are volatile

```
analysis_api_test_db=# \df+ setseed
                                                                                   List of functions
   Schema   |  Name   | Result data type | Argument data types |  Type  | Volatility |  Parallel  |  Owner   | Security | Access privileges | Language | Source code |   Description   
------------+---------+------------------+---------------------+--------+------------+------------+----------+----------+-------------------+----------+-------------+-----------------
 pg_catalog | setseed | void             | double precision    | normal | volatile   | restricted | postgres | invoker  |                   | internal | setseed     | set random seed
(1 row)

analysis_api_test_db=# \df+ random 
                                                                                 List of functions
   Schema   |  Name  | Result data type | Argument data types |  Type  | Volatility |  Parallel  |  Owner   | Security | Access privileges | Language | Source code | Description  
------------+--------+------------------+---------------------+--------+------------+------------+----------+----------+-------------------+----------+-------------+--------------
 pg_catalog | random | double precision |                     | normal | volatile   | restricted | postgres | invoker  |                   | internal | drandom     | random value
```
The issue here is that modifying the seed could be considered as part of the output as that is not cacheable. For performance reasons we could consider setting it as STABLE as done in the quota/redis calls.

rafatower replied:
hm, you're right, a call to `setseed` provokes a side effect (and global state change, namely the random number generator) that prevents it from being parallel safe.

Algunenano commented:
> I'd be nice to copy the volatility and "parallelizability" guidelines in the CONTRIBUTING.md as well here.
Will do

Algunenano commented:
Rebased with fix for bsd-systems compatibility (as https://github.com/CartoDB/cartodb-postgresql/pull/315)

Algunenano commented:
Rebased on top of GWR introduction and added categories to the new functions.

It'd nice if we could merge this and deploy it with GWR so we don't have to do 2 different deploys. cc/ @andy-esch @jgoizueta 
Related to this, a warning has appeared when running python tests: https://github.com/CartoDB/crankshaft/issues/188

Algunenano commented:
After checking the deploy method with @jgoizueta I''ve seen that this projects commits release files (cartodb-postgresql doesn't) so instead of stripping the PARALLEL labels when generating the release file we need to do it in the deploy phase. Should be ok with this last commit.

Also we've decided to do 2 different deploys.

rafatower commented:
:+1:
